File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/util/Tree.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.tinkerpop.gremlin.process.traversal.step.util;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;

/**
 * A tree data structure with a tree-shaped public API.
 * <p>
 * Iteration order over the nodes at a given level is not promised. Because children are keyed by node value,
 * sibling branches that resolve to the same value are represented as a single node; callers that require every
 * path to be preserved should use {@link org.apache.tinkerpop.gremlin.process.traversal.Path} instead.
 *
 * @author Marko A. Rodriguez (http://markorodriguez.com)
 */
public final class Tree<T> implements Serializable {

    private final Map<T, Tree<T>> children = new HashMap<>();

    public Tree() {
    }

    /**
     * Creates a tree with the given values as root keys, each mapping to an empty {@link Tree}.
     */
    @SafeVarargs
    public Tree(final T... rootValues) {
        for (final T t : rootValues) {
            this.children.put(t, new Tree<>());
        }
    }

    // ------------------------------------------------------------------
    // navigation
    // ------------------------------------------------------------------

    /**
     * Returns the set of keys at the root of this tree. The returned set is unmodifiable.
     */
    public Set<T> rootNodes() {
        return Collections.unmodifiableSet(this.children.keySet());
    }

    /**
     * Returns the child subtree for the given key.
     *
     * @throws IllegalArgumentException if no child exists for the given key
     */
    public Tree<T> childAt(final T key) {
        if (!this.children.containsKey(key)) {
            throw new IllegalArgumentException("Tree has no child for key: " + key);
        }
        return this.children.get(key);
    }

    /**
     * Returns {@code true} if the given key is an immediate child of this tree.
     */
    public boolean hasChild(final T key) {
        return this.children.containsKey(key);
    }

    /**
     * Returns {@code true} if the given value appears as a key anywhere in this tree (recursive).
     */
    public boolean contains(final T value) {
        if (this.children.containsKey(value)) {
            return true;
        }
        for (final Tree<T> sub : this.children.values()) {
            if (sub.contains(value)) {
                return true;
            }
        }
        return false;
    }

    /**
     * Recursively searches the tree for the first subtree rooted at {@code key} and returns it. The search visits
     * direct children first and then recurses, but iteration order across siblings is unspecified.
     */
    public Optional<Tree<T>> findSubtree(final T key) {
        if (this.children.containsKey(key)) {
            return Optional.of(this.children.get(key));
        }
        for (final Tree<T> sub : this.children.values()) {
            final Optional<Tree<T>> found = sub.findSubtree(key);
            if (found.isPresent()) {
                return found;
            }
        }
        return Optional.empty();
    }

    /**
     * Returns the existing child for {@code key}, or inserts and returns a new empty {@link Tree} if absent.
     */
    public Tree<T> getOrCreateChild(final T key) {
        Tree<T> child = this.children.get(key);
        if (null == child) {
            child = new Tree<>();
            this.children.put(key, child);
        }
        return child;
    }

    // ------------------------------------------------------------------
    // structural
    // ------------------------------------------------------------------

    /**
     * Returns {@code true} if this tree has no children. An empty tree is considered a leaf.
     */
    public boolean isLeaf() {
        return this.children.isEmpty();
    }

    /**
     * Returns the total number of nodes (keys) in the tree, counted recursively.
     */
    public int nodeCount() {
        int count = this.children.size();
        for (final Tree<T> sub : this.children.values()) {
            count += sub.nodeCount();
        }
        return count;
    }

    /**
     * Returns the keys at the given depth. Depth {@code 0} returns the root keys.
     */
    public List<T> getNodesAtDepth(final int depth) {
        final List<T> list = new ArrayList<>();
        for (final Tree<T> t : this.getTreesAtDepth(depth)) {
            list.addAll(t.children.keySet());
        }
        return list;
    }

    /**
     * Returns the trees at the given depth. Depth {@code 0} returns a singleton list containing this tree.
     * Depths beyond the tree's height return an empty list.
     */
    public List<Tree<T>> getTreesAtDepth(final int depth) {
        if (depth < 0) {
            return Collections.emptyList();
        }
        List<Tree<T>> currentDepth = Collections.singletonList(this);
        for (int i = 0; i < depth; i++) {
            final List<Tree<T>> next = new ArrayList<>();
            for (final Tree<T> t : currentDepth) {
                next.addAll(t.children.values());
            }
            if (next.isEmpty()) {
                return Collections.emptyList();
            }
            currentDepth = next;
        }
        return currentDepth;
    }

    /**
     * Returns all keys whose subtrees are leaves.
     */
    public List<T> getLeafNodes() {
        final List<T> leaves = new ArrayList<>();
        collectLeafKeys(leaves);
        return leaves;
    }

    private void collectLeafKeys(final List<T> out) {
        for (final Map.Entry<T, Tree<T>> entry : this.children.entrySet()) {
            if (entry.getValue().isLeaf()) {
                out.add(entry.getKey());
            } else {
                entry.getValue().collectLeafKeys(out);
            }
        }
    }

    /**
     * Returns single-key trees representing each leaf key in this tree, preserving the original
     * key-to-empty-subtree mapping.
     */
    public List<Tree<T>> getLeafTrees() {
        final List<Tree<T>> leaves = new ArrayList<>();
        collectLeafTrees(leaves);
        return leaves;
    }

    private void collectLeafTrees(final List<Tree<T>> out) {
        for (final Map.Entry<T, Tree<T>> entry : this.children.entrySet()) {
            if (entry.getValue().isLeaf()) {
                final Tree<T> leaf = new Tree<>();
                leaf.children.put(entry.getKey(), entry.getValue());
                out.add(leaf);
            } else {
                entry.getValue().collectLeafTrees(out);
            }
        }
    }

    // ------------------------------------------------------------------
    // composition
    // ------------------------------------------------------------------

    /**
     * Recursively merges {@code other} into this tree. For overlapping keys, child subtrees are merged in turn.
     * For keys present only in {@code other}, the corresponding subtree reference is adopted directly.
     */
    public void addTree(final Tree<T> other) {
        other.children.forEach((k, t) -> {
            if (this.children.containsKey(k)) {
                this.children.get(k).addTree(t);
            } else {
                this.children.put(k, t);
            }
        });
    }

    /**
     * Splits this tree into one tree per root key. If the tree has a single root, returns a singleton list
     * containing this tree.
     */
    public List<Tree<T>> splitParents() {
        if (this.children.size() == 1) {
            return Collections.singletonList(this);
        }
        final List<Tree<T>> parents = new ArrayList<>();
        this.children.forEach((k, t) -> {
            final Tree<T> parentTree = new Tree<>();
            parentTree.children.put(k, t);
            parents.add(parentTree);
        });
        return parents;
    }

    // ------------------------------------------------------------------
    // output
    // ------------------------------------------------------------------

    /**
     * Produces a formatted string representation of the tree structure using a {@code |--} ASCII style.
     */
    public String prettyPrint() {
        final StringBuilder builder = new StringBuilder();
        prettyPrint(builder, "");
        final String pretty = builder.toString();
        return pretty.isEmpty() ? pretty : pretty.substring(0, pretty.length() - System.lineSeparator().length());
    }

    private void prettyPrint(final StringBuilder builder, final String prefix) {
        for (final Map.Entry<T, Tree<T>> entry : this.children.entrySet()) {
            builder.append(prefix).append("|--").append(entry.getKey());
            builder.append(System.lineSeparator());
            entry.getValue().prettyPrint(builder, prefix + "   ");
        }
    }

    // ------------------------------------------------------------------
    // identity
    // ------------------------------------------------------------------

    /**
     * Structural recursive equality. Order across siblings is irrelevant; two trees are equal if they have the
     * same set of keys and each key's subtree is recursively equal.
     */
    @Override
    public boolean equals(final Object other) {
        if (this == other) return true;
        if (!(other instanceof Tree)) return false;
        final Tree<?> that = (Tree<?>) other;
        return this.children.equals(that.children);
    }

    @Override
    public int hashCode() {
        return this.children.hashCode();
    }

    @Override
    public String toString() {
        return this.children.toString();
    }
}

File: gremlin-dotnet/src/Gremlin.Net/Structure/Tree.cs
#region License

/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#endregion

using System;
using System.Collections.Generic;
using System.Text;

namespace Gremlin.Net.Structure
{
    /// <summary>
    ///     A tree data structure with a tree-shaped public API.
    /// </summary>
    /// <remarks>
    ///     Iteration order over the nodes at a given level follows insertion order. Because children are keyed by node
    ///     value, sibling branches that resolve to the same value are represented as a single node; callers that require
    ///     every path to be preserved should use <see cref="Path" /> instead.
    ///     <para>
    ///         Children are stored in an ordered list of (key, child) entries rather than a dictionary so that
    ///         <c>null</c> keys are permitted and key comparisons use value-equality via <see cref="object.Equals(object)" />.
    ///     </para>
    /// </remarks>
    public class Tree : IEquatable<Tree>
    {
        private readonly List<KeyValuePair<object?, Tree>> _entries = new List<KeyValuePair<object?, Tree>>();

        /// <summary>
        ///     Initializes a new empty instance of the <see cref="Tree" /> class.
        /// </summary>
        public Tree()
        {
        }

        private int IndexOfKey(object? key)
        {
            for (var i = 0; i < _entries.Count; i++)
            {
                if (Equals(_entries[i].Key, key))
                {
                    return i;
                }
            }

            return -1;
        }

        // ------------------------------------------------------------------
        // navigation
        // ------------------------------------------------------------------

        /// <summary>
        ///     Returns the keys at the root of this tree in insertion order.
        /// </summary>
        public IReadOnlyList<object?> RootNodes()
        {
            var keys = new List<object?>(_entries.Count);
            foreach (var entry in _entries)
            {
                keys.Add(entry.Key);
            }

            return keys;
        }

        /// <summary>
        ///     Returns the child subtree for the given key.
        /// </summary>
        /// <param name="key">The key whose child subtree should be returned.</param>
        /// <returns>The child <see cref="Tree" /> for the given key.</returns>
        /// <exception cref="KeyNotFoundException">Thrown if no child exists for the given key.</exception>
        public Tree ChildAt(object? key)
        {
            var index = IndexOfKey(key);
            if (index < 0)
            {
                throw new KeyNotFoundException($"Tree has no child for key: {key}");
            }

            return _entries[index].Value;
        }

        /// <summary>
        ///     Returns <c>true</c> if the given key is an immediate child of this tree.
        /// </summary>
        /// <param name="key">The key to look for.</param>
        public bool HasChild(object? key)
        {
            return IndexOfKey(key) >= 0;
        }

        /// <summary>
        ///     Returns <c>true</c> if the given value appears as a key anywhere in this tree (recursive).
        /// </summary>
        /// <param name="value">The value to search for.</param>
        public bool Contains(object? value)
        {
            foreach (var entry in _entries)
            {
                if (Equals(entry.Key, value))
                {
                    return true;
                }

                if (entry.Value.Contains(value))
                {
                    return true;
                }
            }

            return false;
        }

        /// <summary>
        ///     Recursively searches the tree for the first subtree rooted at <paramref name="key" /> and returns it.
        ///     Direct children are visited before recursing.
        /// </summary>
        /// <param name="key">The key to search for.</param>
        /// <returns>The first matching subtree, or <c>null</c> if no match is found.</returns>
        public Tree? FindSubtree(object? key)
        {
            var index = IndexOfKey(key);
            if (index >= 0)
            {
                return _entries[index].Value;
            }

            foreach (var entry in _entries)
            {
                var found = entry.Value.FindSubtree(key);
                if (found != null)
                {
                    return found;
                }
            }

            return null;
        }

        /// <summary>
        ///     Returns the existing child for <paramref name="key" />, or inserts and returns a new empty
        ///     <see cref="Tree" /> if absent.
        /// </summary>
        /// <param name="key">The key whose child should be returned or created.</param>
        public Tree GetOrCreateChild(object? key)
        {
            var index = IndexOfKey(key);
            if (index >= 0)
            {
                return _entries[index].Value;
            }

            var child = new Tree();
            _entries.Add(new KeyValuePair<object?, Tree>(key, child));
            return child;
        }

        // ------------------------------------------------------------------
        // structural
        // ------------------------------------------------------------------

        /// <summary>
        ///     Returns <c>true</c> if this tree has no children. An empty tree is considered a leaf.
        /// </summary>
        public bool IsLeaf()
        {
            return _entries.Count == 0;
        }

        /// <summary>
        ///     Returns the total number of nodes (keys) in the tree, counted recursively.
        /// </summary>
        public int NodeCount()
        {
            var count = _entries.Count;
            foreach (var entry in _entries)
            {
                count += entry.Value.NodeCount();
            }

            return count;
        }

        /// <summary>
        ///     Returns the keys at the given depth. Depth <c>0</c> returns the root keys. Negative depths or depths
        ///     beyond the tree's height return an empty list.
        /// </summary>
        /// <param name="depth">The depth, where <c>0</c> is the root.</param>
        public IReadOnlyList<object?> GetNodesAtDepth(int depth)
        {
            var list = new List<object?>();
            foreach (var tree in GetTreesAtDepth(depth))
            {
                foreach (var entry in tree._entries)
                {
                    list.Add(entry.Key);
                }
            }

            return list;
        }

        /// <summary>
        ///     Returns the trees at the given depth. Depth <c>0</c> returns a singleton list containing this tree.
        ///     Depths beyond the tree's height (or negative depths) return an empty list.
        /// </summary>
        /// <param name="depth">The depth, where <c>0</c> is the root.</param>
        public IReadOnlyList<Tree> GetTreesAtDepth(int depth)
        {
            if (depth < 0)
            {
                return new List<Tree>(0);
            }

            var currentDepth = new List<Tree> { this };
            for (var i = 0; i < depth; i++)
            {
                var next = new List<Tree>();
                foreach (var tree in currentDepth)
                {
                    foreach (var entry in tree._entries)
                    {
                        next.Add(entry.Value);
                    }
                }

                if (next.Count == 0)
                {
                    return new List<Tree>(0);
                }

                currentDepth = next;
            }

            return currentDepth;
        }

        /// <summary>
        ///     Returns all keys whose subtrees are leaves.
        /// </summary>
        public IReadOnlyList<object?> GetLeafNodes()
        {
            var leaves = new List<object?>();
            CollectLeafKeys(leaves);
            return leaves;
        }

        private void CollectLeafKeys(List<object?> output)
        {
            foreach (var entry in _entries)
            {
                if (entry.Value.IsLeaf())
                {
                    output.Add(entry.Key);
                }
                else
                {
                    entry.Value.CollectLeafKeys(output);
                }
            }
        }

        /// <summary>
        ///     Returns single-key trees representing each leaf key in this tree, preserving the original
        ///     key-to-empty-subtree mapping.
        /// </summary>
        public IReadOnlyList<Tree> GetLeafTrees()
        {
            var leaves = new List<Tree>();
            CollectLeafTrees(leaves);
            return leaves;
        }

        private void CollectLeafTrees(List<Tree> output)
        {
            foreach (var entry in _entries)
            {
                if (entry.Value.IsLeaf())
                {
                    var leaf = new Tree();
                    leaf._entries.Add(new KeyValuePair<object?, Tree>(entry.Key, entry.Value));
                    output.Add(leaf);
                }
                else
                {
                    entry.Value.CollectLeafTrees(output);
                }
            }
        }

        // ------------------------------------------------------------------
        // composition
        // ------------------------------------------------------------------

        /// <summary>
        ///     Recursively merges <paramref name="other" /> into this tree. For overlapping keys, child subtrees are
        ///     merged in turn. For keys present only in <paramref name="other" />, the corresponding subtree is adopted
        ///     directly.
        /// </summary>
        /// <param name="other">The tree to merge into this one.</param>
        /// <exception cref="ArgumentNullException">Thrown if <paramref name="other" /> is <c>null</c>.</exception>
        public void AddTree(Tree other)
        {
            if (other == null) throw new ArgumentNullException(nameof(other));

            foreach (var entry in other._entries)
            {
                var index = IndexOfKey(entry.Key);
                if (index >= 0)
                {
                    _entries[index].Value.AddTree(entry.Value);
                }
                else
                {
                    _entries.Add(new KeyValuePair<object?, Tree>(entry.Key, entry.Value));
                }
            }
        }

        /// <summary>
        ///     Splits this tree into one tree per root key. If the tree has a single root, returns a singleton list
        ///     containing this tree.
        /// </summary>
        public IReadOnlyList<Tree> SplitParents()
        {
            if (_entries.Count == 1)
            {
                return new List<Tree> { this };
            }

            var parents = new List<Tree>();
            foreach (var entry in _entries)
            {
                var parentTree = new Tree();
                parentTree._entries.Add(new KeyValuePair<object?, Tree>(entry.Key, entry.Value));
                parents.Add(parentTree);
            }

            return parents;
        }

        // ------------------------------------------------------------------
        // output
        // ------------------------------------------------------------------

        /// <summary>
        ///     Produces a formatted string representation of the tree structure using a <c>|--</c> ASCII style.
        ///     Each level is indented by three spaces relative to its parent, and there is no trailing newline.
        /// </summary>
        public string PrettyPrint()
        {
            const string lineSeparator = "\n";
            var builder = new StringBuilder();
            PrettyPrint(builder, "", lineSeparator);
            var pretty = builder.ToString();
            return pretty.Length == 0
                ? pretty
                : pretty.Substring(0, pretty.Length - lineSeparator.Length);
        }

        private void PrettyPrint(StringBuilder builder, string prefix, string lineSeparator)
        {
            foreach (var entry in _entries)
            {
                builder.Append(prefix).Append("|--").Append(entry.Key);
                builder.Append(lineSeparator);
                entry.Value.PrettyPrint(builder, prefix + "   ", lineSeparator);
            }
        }

        // ------------------------------------------------------------------
        // identity
        // ------------------------------------------------------------------

        /// <summary>
        ///     Structural recursive equality. Order across siblings is irrelevant; two trees are equal if they have the
        ///     same set of keys and each key's subtree is recursively equal.
        /// </summary>
        /// <param name="other">The other tree to compare against.</param>
        public bool Equals(Tree? other)
        {
            if (ReferenceEquals(null, other)) return false;
            if (ReferenceEquals(this, other)) return true;
            if (_entries.Count != other._entries.Count) return false;

            foreach (var entry in _entries)
            {
                var index = other.IndexOfKey(entry.Key);
                if (index < 0) return false;
                if (!entry.Value.Equals(other._entries[index].Value)) return false;
            }

            return true;
        }

        /// <inheritdoc />
        public override bool Equals(object? obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            if (ReferenceEquals(this, obj)) return true;
            if (obj.GetType() != GetType()) return false;
            return Equals((Tree) obj);
        }

        /// <inheritdoc />
        public override int GetHashCode()
        {
            // Order-independent combination of root keys so that equal trees (which ignore sibling order) produce
            // equal hash codes. Subtrees are intentionally not folded in to keep this cheap while remaining
            // content-aware via the root keys.
            unchecked
            {
                var hashCode = 0;
                foreach (var entry in _entries)
                {
                    hashCode += entry.Key?.GetHashCode() ?? 0;
                }

                return hashCode;
            }
        }

        /// <inheritdoc />
        public override string ToString()
        {
            var builder = new StringBuilder();
            builder.Append('{');
            for (var i = 0; i < _entries.Count; i++)
            {
                if (i > 0) builder.Append(", ");
                builder.Append(_entries[i].Key).Append('=').Append(_entries[i].Value);
            }

            builder.Append('}');
            return builder.ToString();
        }
    }
}
